app/server/utils/ussd/token_processor.py

```diff
@@ -0,0 +1,147 @@
+
+    @staticmethod
+    def send_balance_sms(user: User):
+        def get_token_info(transfer_account: TransferAccount):
+            token = transfer_account.token
+            limit = TokenProcessor.get_limit(user, token)
+            return {
+                # TODO: is it name or symbol?
+                "name": token.name,
+                "balance": transfer_account.balance,
+                "limit": limit.transfer_balance_percentage,
+                "exchange_rate": TokenProcessor.get_exchange_rate(user, token)
+            }
+        # TODO: should we merge transfer accounts with same tokens..? any currencies we should filter out?
+        token_info = map(get_token_info, user.transfer_accounts)
+        token_balances = "\n".join(map(lambda x: f"{x['name']} {x['balance']}", token_info))
+        # this one is also not well generalized currently - how get real world currency name of reserve token?
+        token_exchanges = "\n".join(
+            map(lambda x: f"{x['limit'] * x['balance']} {x['name']} (1 {x['name']} = {x['exchange_rate']} KSH)",
+                token_info))
+
+        TokenProcessor.send_sms(user, "send_balance_sms", token_balances=token_balances,
+                                token_exchanges=token_exchanges)
+
+    @staticmethod
+    def fetch_exchange_rate(user: User):
+        from_token = default_token(user)
+
+        limit = TokenProcessor.get_limit(user, from_token)
+        exchange_limit = limit.transfer_balance_percentage * default_transfer_account(user).balance
+        exchange_rate = TokenProcessor.get_exchange_rate(user, from_token)
+
+        TokenProcessor.send_sms(user, "exchange_rate_sms", token_name=from_token.name, exchange_rate=exchange_rate,
+                                exchange_sample_value={exchange_rate * float(1000)}, exchange_limit={exchange_limit},
+                                limit_period={limit.time_period_days})
+
+    @staticmethod
+    def send_token(sender: User, recipient: User, amount: float, reason_str: str, reason_id: int):
+        try:
+            TokenProcessor.transfer_token(sender, recipient, amount, reason_id)
+
+            tx_time = datetime.datetime.now()
+            sender_balance = TokenProcessor.get_balance(sender)
+            recipient_balance = TokenProcessor.get_balance(recipient)
+            TokenProcessor.send_success_sms("send_token_sender_sms", amount, sender, recipient, reason_str, tx_time,
+                                            sender_balance)
+            TokenProcessor.send_success_sms("send_token_recipient_sms", amount, recipient, sender, reason_str, tx_time,
+                                            recipient_balance)
+        except Exception:
+            TokenProcessor.send_sms(sender, "send_token_error_sms", amount=amount,
+                                    token_name=default_token(sender).name, recipient=recipient.user_details())
+
+    @staticmethod
+    def exchange_token(sender: User, agent: User, amount: float):
```

Comment:
i'm a wee bit confused by this method @wangfowen - it creates a credit transfer rather than an exchange and then doesn't seem to do anything with it?